crates/multi-transaction-payment/src/tests.rs

```diff
@@ -0,0 +1,417 @@
+            RuntimeOrigin::signed(lp),
+            currency_1,
+            currency_2,
+            1000000000000000000000,
+            1000000000000000000000,
+            0,
+            0,
+            u64::MAX
+        ));
+    }
+    #[derive(Debug, PartialEq)]
+    pub(super) struct BalanceChange {
+        pub(super) native_cost: i128,
+        pub(super) fee_cost: i128,
+    }
+
+    pub(super) fn dry_run_given_call(
+        who: u128,
+        native_currency: CurrencyId,
+        fee_currency: CurrencyId,
+        pays_fee: Pays,
+        call: RuntimeCall,
+        tip: u128,
+    ) -> BalanceChange {
+        dry_run(|| {
+            let initial_fee_balance = 1000000000000;
+            assert_ok!(Tokens::set_balance(
+                RuntimeOrigin::root(),
+                who,
+                fee_currency,
+                initial_fee_balance,
+                0
+            ));
+            let initial_native_balance = Tokens::free_balance(native_currency, &who);
+
+            let payment_extension = pallet_transaction_payment::ChargeTransactionPayment::<Test>::from(tip);
+
+            let info = call.get_dispatch_info();
+            let len = 1231;
+
+            // check that validate executes successfully, but don't commit storage
+            dry_run(|| {
+                payment_extension.validate(&who, &call, &info, len).unwrap();
+                if let Pays::Yes = pays_fee {
+                    assert!(Tokens::free_balance(fee_currency, &who) < initial_fee_balance);
+                }
+            });
+
+            let pre = payment_extension.pre_dispatch(&who, &call, &info, len).unwrap();
+            if let Pays::Yes = pays_fee {
+                assert!(Tokens::free_balance(fee_currency, &who) < initial_fee_balance);
+            }
+
+            let result = call.dispatch(RuntimeOrigin::signed(who));
+            let post_info = match result {
+                Ok(post_info) => post_info,
+                Err(err) => err.post_info,
+            };
+
+            Payment::post_dispatch(Some(pre), &info, &post_info, len, &Ok(())).unwrap();
+
+            BalanceChange {
+                native_cost: (initial_native_balance as i128) - (Tokens::free_balance(native_currency, &who) as i128),
+                fee_cost: (initial_fee_balance as i128) - (Tokens::free_balance(fee_currency, &who) as i128),
+            }
+        })
+    }
+
+    pub(super) fn dry_run_native(expected_weight: u64, actual_weight: Option<u64>, err: bool, pays_fee: Pays) -> u128 {
+        let who = 1;
+        let native_currency = GetNativeCurrencyId::get();
+
+        let call = RuntimeCall::Testing(testing_helpers::Call::weighted {
+            expected: Weight::from_ref_time(expected_weight),
+            actual_weight: actual_weight.map(|x| Weight::from_ref_time(x)),
+            err,
+            pays_fee,
+        });
+        dry_run_given_call(who, native_currency, native_currency, pays_fee, call, 0).native_cost as u128
```

Comment:
This should always be less than or equal to zero, should the return type should be signed

Reply:
This is dry-running a call using the native currency as fee, i.e. not swap is being performed. So the cost will be 0 or positive